File: java/40-Combination-Sum-II.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        List<Integer> ls = new ArrayList<Integer>();
        comb(candidates, target, ans, ls, 0);
        return ans;
    }
    
    public void comb(int[] candidates, int target, List<List<Integer>> ans, List<Integer> ls, int index) {
        if (target == 0) {
            ans.add(new ArrayList(ls));
        }
        else if (target<0) 
            return;
        else {
            for (int i = index; i<candidates.length; i++) {
                if(i>index && candidates[i] == candidates[i-1]) continue;
                ls.add(candidates[i]);
                comb(candidates, target-candidates[i], ans, ls, i+1);
                ls.remove(ls.get(ls.size()-1));
            }
        }
    }
}

File: javascript/40-Combination-Sum-II.js
//////////////////////////////////////////////////////////////////////////////
// Backtracking
// Time: Theta(2^log(n)) O(2^n)
// Space: Theta(2^log(n)) O(2^n)
//////////////////////////////////////////////////////////////////////////////

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
function combinationSum2(candidates, target) {
  candidates.sort((a, b) => a - b);

  const combos = [];
  const combo = [];
  const map = Object.create(null);

  for (let i = 0; i < candidates.length; ++i) {
    map[candidates[i]] = i;
  }

  getCombos(target);
  return combos;

  /**
   * @param {number} target
   * @param {number=} start = `0`
   * @return {void}
   */
  function getCombos(target, start = 0) {
    if (target in map && start <= map[target]) {
      combo.push(target);
      combos.push(combo.slice());
      combo.pop();
    }

    const mid = Math.floor(target / 2);
    for (let i = start; i < candidates.length && candidates[i] <= mid; ++i) {
      if (i !== start && candidates[i] === candidates[i - 1]) {
        continue;
      }
      combo.push(candidates[i]);
      getCombos(target - candidates[i], i + 1);
      combo.pop();
    }
  }
}
